src/sinks/loki/event.rs

```diff
@@ -56,23 +65,45 @@ impl Encoder<Vec<LokiRecord>> for LokiBatchEncoder {
 
 #[derive(Debug, Default, Serialize)]
 pub struct LokiBatch {
-    stream: HashMap<String, String>,
-    values: Vec<LokiEvent>,
+    stream_by_labels: HashMap<String, LokiStream>,
     #[serde(skip)]
     finalizers: EventFinalizers,
 }
 
+#[derive(Debug, Default, Serialize)]
+pub struct LokiStream {
+    stream: HashMap<String, String>,
+    values: Vec<LokiEvent>,
+}
+
 impl From<Vec<LokiRecord>> for LokiBatch {
     fn from(events: Vec<LokiRecord>) -> Self {
         let mut result = events
             .into_iter()
             .fold(Self::default(), |mut res, mut item| {
                 res.finalizers.merge(item.take_finalizers());
-                res.stream.extend(item.labels.into_iter());
-                res.values.push(item.event);
+                item.labels.sort();
+                let labels: String = item
+                    .labels
+                    .iter()
+                    .flat_map(|(a, b)| [a, "→", b, "∇"])
```

Comment:
Note: I'm not entirely clear why we picked these separators initially, and it might make sense to update them to `"` and `, ` as to match the [promtail](https://github.com/grafana/loki/blob/a4f306399ad39aef8af321b527d90790aa02ec39/clients/pkg/promtail/client/batch.go) behavior.

I'll probably follow up after on this PR to see if it makes a meaningful difference.

Reply:
It appears that the stringification of labels here is not something that is visible to users. Therefore, we have flexibility in how we choose to insert separators. However, upon further consideration, the previous approach -- using separators that are unlikely to appear in keys or values -- may not be a safe one, as these separators could potentially appear in keys or values. In the latest commit, I have implemented a solution that joins keys and values with commas and escapes any commas that appear in keys or values. This approach should address the issue of separators potentially appearing in keys or values.